src/main/java/io/vertx/core/http/impl/Http1xPool.java

```diff
@@ -81,39 +83,40 @@ public HttpClientStream createStream(ClientConnection conn) {
     return conn;
   }
 
-  // Called when the request has ended
   public void recycle(ClientConnection conn) {
-    if (pipelining) {
-      doRecycle(conn);
-    }
-  }
-
-  // Called when the response has ended
-  public void responseEnded(ClientConnection conn, boolean close) {
-    if ((pipelining || keepAlive) && !close && conn.getCurrentRequest() == null) {
-      doRecycle(conn);
-    } else {
-      // Close it now
-      conn.close();
-    }
-  }
-
-  private void doRecycle(ClientConnection conn) {
     synchronized (queue) {
       Waiter waiter = queue.getNextWaiter();
       if (waiter != null) {
-        Context context = waiter.context;
```

Comment:
the important thing is that there should be no runOnContext between the time a connection was check to be reusable because pipelining limit is not reached and the time it is dispatched (where the getOutStandingRequestCount() will incrase). Otherwise two checks be valid and a number > pipeliningLimit of requests can be done (race condition)